scripts: tests: sanitycheck: Add testcase for add_testcases function of testsuite class

aasthagr opened the pull request:
…ion of testsuite class

test_testsuite_class.py: Add testcases for add_testcases function of testsuite class in sanitycheck.
test_sanitycheck.py: Add testcases for correct & incorrect testcase/platform yaml schemas. Moved the fixture to conftest.py
test_data/testcases/tests & /samples : Testcase root directory to add all the testcases & to test add_testcases function.
conftest.py: Module for common pytest fixtures, also used for passing data from one testcase to another.
Note: conftest.py has a class_testsuite fixture where board_root is defined as the directory which will be added in a separate PR.

Signed-off-by: Aastha Grover <aastha.grover@intel.com>

aasthagr committed:
sanitycheck: add testcases for add_testcases function of testsuite class

test_testsuite_class.py: Add testcases for add_testcases function
of testsuite class in sanitycheck.
test_data/testcases/tests & /samples : Testcase root directory
to add all the testcases & to test add_testcases function.
conftest.py: Module for common pytest fixtures, also used for
passing data from one testcase to another.
Note: conftest.py has a class_testsuite fixture where board_root is
defined as the directory which will be added in a separate PR.

Signed-off-by: Aastha Grover <aastha.grover@intel.com>

nashif commented on scripts/tests/sanitycheck/test_testsuite_class.py:
```diff
@@ -0,0 +1,49 @@
+                          'scripts/tests/sanitycheck/test_data/testcases/tests/test_a/kernel.fifo',
+                          'scripts/tests/sanitycheck/test_data/testcases/tests/test_a/kernel.fifo.poll',
+                          'scripts/tests/sanitycheck/test_data/testcases/samples/test_app/sample.app_dev.external_lib']
+    assert set(class_testsuite.testcases.keys()) == set(expected_testcases)
```
why using set() here? This will only hide problems, ie. if there is for whatever reason a duplicate, the set() call will just remove it and still match, just used sorted()

aasthagr replied:
Fixed

nashif commented on scripts/tests/sanitycheck/test_testsuite_class.py:
```diff
@@ -0,0 +1,49 @@
+    class_testsuite.SAMPLE_FILENAME = 'test_sample_app.yaml'
+    class_testsuite.TESTCASE_FILENAME = 'test_data.yaml'
+    class_testsuite.add_testcases()
+    expected_testcases = ['scripts/tests/sanitycheck/test_data/testcases/tests/test_b/kernel.fifo.timeout',
```
You could test against something else, for example:

```
    expected_testcases = [
        'kernel.fifo',
        'kernel.fifo.poll',
        'kernel.fifo.usage',
        'kernel.fifo.usage.poll',
        'kernel.fifo.timeout',
        'kernel.fifo.timeout.poll',
        'sample.app_dev.external_lib'
    ]
    got = [t.id for t in class_testsuite.testcases.values()]
    assert sorted(got) == sorted(expected_testcases)
```

or you could define the common name of the path once and just use the variable everywhere...

It is not fun to review and see those paths, difficult to read and not really needed.

aasthagr replied:
I have cleaned this thing to just compare the last part of the testcases, instead of prefixing a string. Add_testcases function forms a dictionary of testcases where the key is testcase name coming from get_unique function (Which is the relative path to the testcase). But for making the paths look nicer, i am now just comparing the last part of the testcase name.

nashif commented on scripts/tests/sanitycheck/test_testsuite_class.py:
```diff
@@ -0,0 +1,49 @@
+    # Test 2 : Assert Testcase name is expected & all the testcases values are testcase class objects
+    testcase = class_testsuite.testcases.get('scripts/tests/sanitycheck/test_data/testcases/tests/test_a/kernel.fifo')
+    assert testcase.name == 'scripts/tests/sanitycheck/test_data/testcases/tests/test_a/kernel.fifo'
+    assert all(isinstance(n, TestCase) for n in class_testsuite.testcases.values())
```
is there any chance we might not be a testcase object? you have been treating this as such in the previous tests.

aasthagr replied:
There is an edge case in add_testcases function where an error could be raised "Cant load the testcase.yaml file", which I wasn't able to reproduce. So just wanted to check that no error is raised  from the try block. Thats why checking for the end result ie.. testcases dictionary keys & values.

nashif commented on scripts/tests/sanitycheck/test_testsuite_class.py:
```diff
@@ -0,0 +1,49 @@
+                                       'skip', 'platform_exclude', 'platform_whitelist', 'toolchain_exclude', 'toolchain_whitelist',
+                                       'tc_filter', 'timeout', 'harness', 'harness_config', 'build_only', 'build_on_all', 'slow',
+                                       'min_ram', 'depends_on', 'min_flash', 'extra_sections']
+        assert set(expected_variables_to_check) == set(list_of_vars)
```
Not how is this a test, I would rather see those tested when used, and not just list all instances variables and see if they are there.

aasthagr replied:
Removed this test scenario.

aasthagr commented on scripts/tests/sanitycheck/test_data/testcase_incorrect_schema.yaml:
```diff
@@ -0,0 +1,6 @@
+test:
```
The purpose of this yaml file is to test that the error is raised if the testcase yaml files or platform yaml files are not following the expected schema pattern.  I can remove testcase_correct_schema.yaml file and instead use test_data/testcases/test_a/test_data.yaml file for checking the correct testcase schema scenario. But we need incorrect yaml files for testing incorrect schema's to assert that appropriate error is raised. Also we need yaml files for testing platform correct/incorrect schema.

nashif replied:
you can add yaml files for this type of test, but do not just pull them from the real tests, you can create your own for both correct and incorrect behavior.

aasthagr commented on scripts/tests/sanitycheck/test_testsuite_class.py:
```diff
@@ -0,0 +1,49 @@
+def test_testsuite_add_testcases(class_testsuite):
+    """ Testing add_testcase function of Testsuite class in sanitycheck """
+    # Test 1: Check the list of testcases after calling add testcases function is as expected
+    class_testsuite.SAMPLE_FILENAME = 'test_sample_app.yaml'
```
This changed was merged yesterday and I have included those changes in this PR. 

For the other issue,  I will open a separate task to look into that.

aasthagr commented on scripts/tests/sanitycheck/test_sanitycheck.py:
```diff
@@ -16,22 +16,32 @@
+                         [("testcase_correct_schema.yaml", "testcase-schema.yaml"),
+                          ("platform_correct_schema.yaml", "platform-schema.yaml")])
+def test_correct_schema(filename, schema, test_data):
     """ Test to validate the testcase schema"""
```
Looking at _yaml_validate function in sanity_chk/scl.py. There should be an error raised if the schema is not correct and I am asserting that for correct & incorrect scenario. And the code for validating the schema already exists, so indirectly I am validating the schema if that makes sense.

nashif commented on scripts/tests/sanitycheck/test_data/platform_correct_schema.yaml:
```diff
@@ -0,0 +1,2 @@
+testing:
+  ignore_tags: [correct_schema_1,correct_schema_2]
```
just noticed this, why do you have the [] around the list here?

nashif commented on scripts/tests/sanitycheck/test_data/platform_incorrect_schema.yaml:
```diff
@@ -0,0 +1,2 @@
+test:
+  ignore_tags: [incorrect_platform_schema1,incorrect_schema_2]
```
same

nashif commented on scripts/tests/sanitycheck/test_sanitycheck.py:
```diff
@@ -16,22 +16,32 @@
+                          ("platform_correct_schema.yaml", "platform-schema.yaml")])
+def test_correct_schema(filename, schema, test_data):
     """ Test to validate the testcase schema"""
-    filename = test_data + 'testcase_correct_schema.yaml'
```
can you address this comment please.

aasthagr replied:
Sure, Added a separate commit to this PR for schema validation testcases.

aasthagr committed:
scripts: tests: sanitycheck: Add testcases for schema validation.

test_sanitycheck.py: Add testcases for correct & incorrect
testcase/platform yaml schemas.

Signed-off-by: Aastha Grover <aastha.grover@intel.com>
